Migrate Panel and related components

Chng-Zhi-Xuan opened the pull request:
What is the purpose of this pull request? (put "X" next to an item, remove the rest)

• [X] Other, migration

<!--
    If this pull request is addressing an issue, link to the issue: "Fixes #xxx" or "Resolves #xxx"
-->

Part of MarkBind/markbind#266

<!--
    Please ensure your pull request is ready:
    - Bug fix PR that is non-trivial should add a page or unit test for regression testing.
    - Feature PR must add a page to the user guide for demo.
    - Enhancement PR should update the user guide.

    Otherwise, prefix your PR title with "[WIP]".
-->

What changes did you make? (Give an overview)

Global changes |
--- |
Change `v-el` / `v-ref` to **ref** |
Refactor `coerce` in `props` |
`attached()` / `ready()` to **mounted()** with **this.$nextTick()**
Refactor `event names` to be **kebab-case**, [reason here](https://vuejs.org/v2/guide/components-custom-events.html)
Refactor `var` to `let`, (ES6)

Component | Specific changes
--- | --- 
Morph.vue | (Deprecated)<br>Global changes only
Retriever.vue | Change `this.$dispatch` to directly mount retrieved html into retriever contents
PanelSwitch.vue | Change `this.$dispatch` to `this.$emit`
DynamicPanel.vue | (Deprecated)<br>Global changes only
Panel.vue | 1. Remove `events` and `expandCollapseHandler` (not needed)<br>2. Add additional `load-all` attribute to enable non-dynamic behaviour (user request)
Index | Remove deprecated `$broadcast` calls within `installEvents`

Testing instructions:
1. Download [PanelTestV2.zip](https://github.com/MarkBind/vue-strap/files/2117858/PanelTestV2.zip)
2. Extract and run `markbind serve` within folder (after copy pasting updated vue.min.js etc)
3. Test provided panels

yamgent commented on src/DynamicPanel.vue:
```diff
@@ -1,12 +1,12 @@
 <template>
   <panel :header="header" :is-open="isOpen" :type="type" expandable no-switch>
```
Should be `isOpenBool`

yamgent commented on src/Morph.vue:
```diff
@@ -43,14 +43,18 @@
-        coerce: coerce.boolean,
         default: false
       },
       src: {
```
```html
<morph title="Abc" src="Doge/index.md">
</morph>
```

This does not work on the migrated version.

yamgent commented on src/Panel.vue:
```diff
@@ -71,48 +72,43 @@
         type: String,
         default: null
       },
       expandable: {
```
'expandable === false' scenario: The panel should have been expanded when the page loaded.

yamgent commented on src/Panel.vue:
```diff
@@ -160,92 +176,78 @@
     created () {
-      if (this.expanded === null) {
-        this.expanded = !this.canCollapse
+      if (this.expandedBool === null) {
```
This will never evaluate to true, because `expandedBool()` will never be null.

Using the raw variable `this.expanded` should be fine here.

Chng-Zhi-Xuan replied:
Since I'm setting `this.localExpanded` below, I can remove the if statement block entirely.

yamgent commented on src/Panel.vue:
```diff
@@ -124,30 +120,50 @@
-    },
     computed: {
+      // Vue 2.0 coerce migration
+      expandableBool: function () {
```
Doing `expandableBool() {` is sufficient, same for the rest of the migrated coerce methods.

Chng-Zhi-Xuan commented:
**Update**
- Deprecated `Morph` and `DynamicPanel`  after team discussion. (both have been migrated but not used) 
- Add `load-all` attribute to `Panel` to have **non-dynamic** behaviour when `src` is provided
- Refactor local prop copies back to using actual props (works after explicitly assigning computed value to it during `created()`)

**PR**
- Changed description
- Updated `PanelTest.zip` with `v2` version

yamgent commented on src/Panel.vue:
```diff
@@ -160,92 +176,74 @@
         }
       }
+      
+      //Set prop data to computed prop value
```
Space after `//`

yamgent commented on src/index.js:
```diff
@@ -73,13 +69,10 @@ function install (Vue) {
-  })
-  vm.$on('retriever:fetched', function (el) {
-    vm.$compile(el);
+  //  this.$broadcast('modal:show', name)
```
I don't think we should comment this out, the migration helper will not be able to catch these if we forget to remove it subsequently. We should remove them in the `Modal` and `Trigger`'s migration PR instead.

Chng-Zhi-Xuan commented on src/Panel.vue:
```diff
@@ -14,23 +14,23 @@
-                        <panel-switch v-show="canCollapse && !noSwitch && !showCaret" v-bind:is-open="expanded"
-                                      @click.stop.prevent="expand()"></panel-switch>
+                        <panel-switch v-show="canCollapse && !noSwitchBool && !showCaret" v-bind:is-open="expandedBool"
+                                      @click.native.stop.prevent="expand()"
```
Without the `.native` modifier, it doesn't work. The behaviour for `v-on` changed, [documentation here](https://vuejs.org/v2/guide/migration.html#Listening-for-Native-Events-on-Components-with-v-on-changed)

yamgent commented on src/Panel.vue:
```diff
@@ -236,9 +236,10 @@
+      // Edge case where user might want non-expandable panel that isn't expanded by default
+      const notExpandableNoExpand = !this.expandableBool && this.expanded !== 'false';
+      // Set prop data to computed prop value
+      this.expanded =  this.expandedBool || notExpandableNoExpand;
```
A note here for archival purpose: The order of the boolean is creating the "minimized" bug (whereby if the panel is minimized at the start, one must click the button twice to expand the panel).

It should have been:
```
this.expanded = notExpandableNoExpand || this.expandedBool;
```

Alternatively, fix `toBoolean()` to deal with null values (`this.expandedBool` is being evaluated to `null`).

Chng-Zhi-Xuan commented:
**Update**
- Rebase to latest migration
- Fix `this.expanded` Boolean ordering causing **double click minimise bug**
